state_tracker.go

```diff
@@ -137,31 +176,93 @@ func (s *StateTracker) UpdateLastResumableBinlogPositionForTargetVerifier(pos my
 	s.lastStoredBinlogPositionForTargetVerifier = pos
 }
 
-func (s *StateTracker) UpdateLastSuccessfulPaginationKey(table string, paginationKey uint64) {
+func (s *StateTracker) RegisterBatch(table string, index uint64, startPaginationKey uint64, endPaginationKey uint64) {
+	s.CopyRWMutex.Lock()
+	defer s.CopyRWMutex.Unlock()
+
+	if _, ok := s.batchProgress[table]; !ok {
+		s.batchProgress[table] = make(map[uint64]*BatchProgress)
+	}
+
+	s.batchProgress[table][index] = &BatchProgress{
+		StartPaginationKey:  startPaginationKey,
+		EndPaginationKey:    endPaginationKey,
+		LatestPaginationKey: startPaginationKey,
+		Completed:           false,
+	}
+}
+
+func (s *StateTracker) UpdateBatchPosition(table string, index uint64, latestPaginationKey uint64) {
+	s.CopyRWMutex.Lock()
+
+	logger := s.logger.WithField("table", table).WithField("batchId", index)
+
+	_, tableFound := s.batchProgress[table]
+	if !tableFound {
+		logger.Error("tried to mark non-existing batch as completed")
+		return
+	}
+
+	batch, batchFound := s.batchProgress[table][index]
+	if !batchFound {
+		logger.Error("tried to mark non-existing batch as completed")
+		return
+	}
+
+	newLatestPaginationKey := uint64(math.Min(float64(latestPaginationKey), float64(batch.EndPaginationKey)))
+	s.updateSpeedLog(newLatestPaginationKey - batch.LatestPaginationKey)
+
+	// Set latest key to EndPaginationKey if we are ahead due to a fragmented PK
+	batch.LatestPaginationKey = newLatestPaginationKey
+
+	if batch.EndPaginationKey == batch.LatestPaginationKey {
+		logger.Debug("marking batch as completed")
+		batch.Completed = true
+		batch.completedPercentage = 100
+		s.CopyRWMutex.Unlock()
+
+		if s.areAllBatchesCompleted(table) {
+			logger.Info("marking table as done")
+			s.MarkTableAsCompleted(table)
+		}
+	} else {
+		batch.completedPercentage = 100 * (batch.LatestPaginationKey - batch.StartPaginationKey) / (batch.EndPaginationKey - batch.StartPaginationKey)
+		s.CopyRWMutex.Unlock()
+	}
+}
+
+func (s *StateTracker) IsBatchComplete(table string, index uint64) bool {
 	s.CopyRWMutex.Lock()
 	defer s.CopyRWMutex.Unlock()
 
-	deltaPaginationKey := paginationKey - s.lastSuccessfulPaginationKeys[table]
-	s.lastSuccessfulPaginationKeys[table] = paginationKey
+	if _, ok := s.batchProgress[table]; !ok {
+		return false
+	}
+
+	batch, ok := s.batchProgress[table][index]
+	if !ok {
+		return false
+	}
 
-	s.updateSpeedLog(deltaPaginationKey)
+	return batch.Completed
 }
 
-func (s *StateTracker) LastSuccessfulPaginationKey(table string) uint64 {
-	s.CopyRWMutex.RLock()
-	defer s.CopyRWMutex.RUnlock()
+func (s *StateTracker) areAllBatchesCompleted(table string) bool {
+	s.CopyRWMutex.Lock()
```

Comment:
why the change to `Lock`? Aren't we only reading in this method?

Reply:
You are correct, it wrote some commits ago but forgot to change it back. Fixing.